File: src/calc_engine/src/services/copilot/file_manager.py
import os
from google import genai
from fastapi import UploadFile
import tempfile
import time

# Configure API (New Client)
client = None
if "GOOGLE_API_KEY" in os.environ:
    client = genai.Client(api_key=os.environ["GOOGLE_API_KEY"])

class FileManager:
    @staticmethod
    async def upload_file(file: UploadFile, display_name: str = None):
        """
        Uploads a file to Gemini File API using the new SDK.
        """
        if not client:
            raise Exception("GOOGLE_API_KEY not set")

        try:
            # Create temp file because SDK expects a path (usually)
            suffix = os.path.splitext(file.filename)[1]
            with tempfile.NamedTemporaryFile(delete=False, suffix=suffix) as tmp:
                content = await file.read()
                tmp.write(content)
                tmp_path = tmp.name

            try:
                print(f"[FileManager] Uploading {file.filename} ({len(content)} bytes)...")
                
                # New SDK Upload
                uploaded_file = client.files.upload(
                    path=tmp_path,
                    config={"display_name": display_name or file.filename, "mime_type": file.content_type}
                )
                
                print(f"[FileManager] Uploaded: {uploaded_file.uri}")
                
                # Check state for video/audio processing
                FileManager._wait_for_active(uploaded_file.name)
                
                return {
                    "fileUri": uploaded_file.uri,
                    "mimeType": uploaded_file.mime_type,
                    "name": uploaded_file.name
                }
            finally:
                # Cleanup temp
                if os.path.exists(tmp_path):
                    os.unlink(tmp_path)

        except Exception as e:
            print(f"[FileManager] Error: {e}")
            raise e

    @staticmethod
    def _wait_for_active(file_name):
        """Waits for file to be active (processed)"""
        print(f"[FileManager] Waiting for processing: {file_name}")
        max_retries = 30
        for _ in range(max_retries):
            # client.files.get(name=...)
            updated_file = client.files.get(name=file_name)
            if updated_file.state == "ACTIVE":
                print(f"[FileManager] File is ACTIVE: {file_name}")
                return
            elif updated_file.state == "FAILED":
                raise Exception("File processing failed")
            time.sleep(2)
        print("[FileManager] Warning: Timed out waiting for processing")

File: src/backend/src/agents/copilot-domain/file-manager.js
const { GoogleAIFileManager } = require("@google/generative-ai/server");

/**
 * Service to manage file uploads to Gemini File API using the official SDK.
 * Used for large files (Video, Audio) that exceed inline Base64 limits.
 */
class FileManager {
    constructor() {
        this._client = null; // Lazy init
    }

    get client() {
        if (this._client) return this._client;
        if (!process.env.GOOGLE_API_KEY) {
            throw new Error('GOOGLE_API_KEY not set (required for File API). Configure in backend .env');
        }
        this._client = new GoogleAIFileManager(process.env.GOOGLE_API_KEY);
        return this._client;
    }

    /**
     * Uploads a file buffer to Gemini.
     * @param {Buffer} buffer - The file buffer
     * @param {string} mimeType - The file mime type
     * @param {string} displayName - Optional display name
     * @returns {Promise<Object>} - { fileUri, mimeType, name }
     */
    async uploadFile(buffer, mimeType, displayName) {
        try {
            const fs = require('fs');
            const path = require('path');
            const os = require('os');

            const tempDir = os.tmpdir();
            const tempFileWithExt = path.join(tempDir, `upload_${Date.now()}_${Math.random().toString(36).substring(7)}.${mimeType.split('/')[1] || 'bin'}`);

            // Write buffer to temp file
            fs.writeFileSync(tempFileWithExt, buffer);

            console.log(`[FileManager] Uploading ${displayName || 'file'} to Gemini...`);

            const uploadResponse = await this.client.uploadFile(tempFileWithExt, {
                mimeType: mimeType,
                displayName: displayName || "Uploaded File"
            });

            console.log(`[FileManager] Upload successful: ${uploadResponse.file.uri}`);

            // Cleanup temp file
            fs.unlinkSync(tempFileWithExt);

            return {
                fileUri: uploadResponse.file.uri,
                mimeType: uploadResponse.file.mimeType,
                name: uploadResponse.file.name
            };

        } catch (error) {
            console.error('[FileManager] Upload failed:', error);
            throw error;
        }
    }

    /**
     * Gets file state.
     * @param {string} name - The file resource name (e.g. files/...)
     */
    async getFile(name) {
        return await this.client.getFile(name);
    }
}

module.exports = new FileManager();
